File: GildedRoseKata/scala/src/test/scala/com/gildedrose/Generators.scala
package com.gildedrose

import org.scalacheck.Gen

object Generators {

  val daysGen: Gen[Int] = Gen.posNum[Int]

  def itemGen(name: String): Gen[Item] = for {
    s <- Gen.posNum[Int]
    q <- Gen.choose(0, 50)
  } yield new Item(name, s, q)

  val sulfurasGen: Gen[Item] = itemGen(Sulfuras.name)
  val agedBrieGen: Gen[Item] = itemGen(AgedBrie.name)
  val elixirGen: Gen[Item] = itemGen("Elixir of the Mongoose")
  val dexterityGen: Gen[Item] = itemGen("+5 Dexterity Vest")

  val backstagePassesFarGen: Gen[(Int, Item)] = for {
    s <- Gen.choose(10, 1000)
    d <- Gen.choose(0, s - 10)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(BackstagePasses.name, s, q))

  val backstagePassesCloseGen: Gen[(Int, Item)] = for {
    s <- Gen.choose(5, 10)
    d <- Gen.choose(0, s - 5)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(BackstagePasses.name, s, q))

  val backstagePassesClosestGen: Gen[(Int, Item)] = for {
    s <- Gen.choose(0, 5)
    d <- Gen.choose(0, s)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(BackstagePasses.name, s, q))

  val backstagePassesWillExpireGen: Gen[(Int, Item)] = for {
    d <- Gen.choose(6, 1000)
    s <- Gen.choose(0, 5)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(BackstagePasses.name, s, q))

  val conjuredWillExpireGen: Gen[(Int, Item)] = for {
    d <- Gen.choose(6, 1000)
    s <- Gen.choose(0, 5)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(Conjured.name, s, q))

  val conjuredNotExpiredGen: Gen[(Int, Item)] = for {
    s <- Gen.choose(6, 1000)
    d <- Gen.choose(0, 5)
    q <- Gen.choose(0, 50)
  } yield (d, new Item(Conjured.name, s, q))

  val conjuredGen: Gen[Item] = Gen.oneOf(
    conjuredWillExpireGen.map(_._2),
    conjuredNotExpiredGen.map(_._2),
  )

  val backstagePassesGen: Gen[Item] = Gen.oneOf(
    backstagePassesFarGen.map(_._2),
    backstagePassesCloseGen.map(_._2),
    backstagePassesClosestGen.map(_._2),
    backstagePassesWillExpireGen.map(_._2)
  )

  val itemGen: Gen[Item] = Gen.oneOf(sulfurasGen, agedBrieGen, elixirGen, dexterityGen, backstagePassesGen, conjuredGen)
  val standardItemGen: Gen[Item] = Gen.oneOf(elixirGen, dexterityGen)
}

File: GildedRoseKata/haskell/test/Generators.hs
module Generators  where

import Test.QuickCheck
import GildedRose
import Test.QuickCheck.Instances.Tuple

itemGen :: String -> Bool -> Gen (Item, Positive Int)
itemGen name sellInExpired = do
  days <- (arbitrary :: Gen (Positive Int))
  negativeSellIn <- (arbitrary :: Gen (Negative Int))
  positiveSellIn <- (arbitrary :: Gen (Positive Int))
  let sellIn = if sellInExpired then (getNegative negativeSellIn) else (getPositive positiveSellIn)
  quality <- choose (0, 50)
  return ((Item name sellIn quality), days)

getQuality :: Item -> Int
getQuality (Item _ _ q) = q

getSellIn :: Item -> Int
getSellIn (Item _ s _) = s

backstagePassesGen :: (Int -> (Int, Int)) -> Gen (Item, Positive Int)
backstagePassesGen sellInBoundsFunc = do
  days <- (arbitrary :: Gen (Positive Int))
  sellIn <- choose $ sellInBoundsFunc (getPositive days)
  quality <- choose (0, 50)
  return ((Item "Backstage passes to a TAFKAL80ETC concert" sellIn quality), days)

backstagePassesFarGen :: Gen (Item, Positive Int)
backstagePassesFarGen = backstagePassesGen $ \days -> (days + 10, maxBound :: Int)

backstagePassesCloseGen :: Gen (Item, Positive Int)
backstagePassesCloseGen = backstagePassesGen $ \days -> (days + 5, days + 10)

backstagePassesClosestGen :: Gen (Item, Positive Int)
backstagePassesClosestGen = backstagePassesGen $ \days -> (days, days + 5)

backstagePassesWillExpireGen :: Gen (Item, Positive Int)
backstagePassesWillExpireGen = do
  days <- choose $ (6, 1000)
  sellIn <- choose $ (0, 5)
  quality <- choose (0, 50)
  return ((Item "Backstage passes to a TAFKAL80ETC concert" sellIn quality), Positive days)

dexterityGen :: Bool -> Gen (Item, Positive Int)
dexterityGen = itemGen "+5 Dexterity Vest"

elixirGen :: Bool -> Gen (Item, Positive Int)
elixirGen = itemGen "Elixir of the Mongoose"

agedBrieGen :: Gen (Item, Positive Int)
agedBrieGen = itemGen "Aged Brie" False

sulfurasGen :: Gen (Item, Positive Int)
sulfurasGen = itemGen "Sulfuras, Hand of Ragnaros" False

allItemGen :: Gen Item
allItemGen = oneof $ fmap (fmap fst) [
  dexterityGen False,
    elixirGen False,
    agedBrieGen,
    sulfurasGen,
    (backstagePassesGen (\_ -> (0, 1000)))
  ]

allItemGenExpired :: Gen Item
allItemGenExpired = oneof $ fmap (fmap fst) [
  dexterityGen True,
    elixirGen True
  ]

allItemsGen :: Gen Item -> Gen ([Item], Positive Int)
allItemsGen x = (listOf x) >*< (arbitrary :: Gen (Positive Int))
